gbn/gbn_server.go

```diff
@@ -57,20 +59,38 @@ func (g *GoBackNConn) serverHandshake() error {
 				errChan <- err
 				return
 			}
-			recvChan <- b
+
+			select {
+			case <-g.ctx.Done():
+				return
+			case <-g.quit:
+				return
+			case <-handshakeComplete:
+				return
+			case recvChan <- b:
+			}
 		}
 	}()
 
 	var n uint8
 
 	for {
 		log.Debugf("Waiting for client SYN")
-		recvNext <- 1
+		select {
+		case <-g.ctx.Done():
+			return nil
+		case <-g.quit:
+			return nil
+		case recvNext <- 1:
+		default:
```

Comment:
i dont think it should be buffered cause then we might end up handing on the recv func when we restart here cause it will try recieve more than it should and it might recv something that is meant for the data layer after the handshake.